ichnaea/service/locate.py

```diff
@@ -122,496 +100,536 @@ def map_data(data):
     return mapped
 
 
-def geoip_and_best_guess_country_codes(cell_keys, api_name, client_addr,
-                                       geoip_db, stats_client):
-    """
-    Return (geoip, alpha2) where geoip is the result of a GeoIP lookup
-    and alpha2 is a best-guess ISO 3166 alpha2 country code. The country
-    code guess uses both GeoIP and cell MCCs, preferring GeoIP. Return
-    None for either field if no data is available.
-    """
-    geoip = None
-
-    if client_addr and geoip_db is not None:
-        geoip = geoip_db.geoip_lookup(client_addr)
+class StatsLogger(object):
+
+    def __init__(self, api_key_name, api_key_log, api_name):
+        self.api_key_name = api_key_name
+        self.api_key_log = api_key_log
+        self.api_name = api_name
+        self.stats_client = get_stats_client()
+
+    def stat_count(self, stat):
+        self.stats_client.incr('{api}.{stat}'.format(
+            api=self.api_name, stat=stat))
+
+    def stat_time(self, stat, count):
+        self.stats_client.timing('{api}.{stat}'.format(
+            api=self.api_name, stat=stat), count)
+
+    def log_used(self):
+        if self.api_key_log:
+            self.stat_count('api_log.{key}.{metric}_hit'.format(
+                key=self.api_key_name, metric=self.log_name))
+
+    def log_unused(self):
+        if self.api_key_log:
+            self.stat_count('api_log.{key}.{metric}_miss'.format(
+                key=self.api_key_name, metric=self.log_name))
+
+
+class AbstractLocationProvider(StatsLogger):
+
+    def __init__(self, session, *args, **kwargs):
+        self.session = session
+        self.heka_client = get_heka_client()
+        super(AbstractLocationProvider, self).__init__(*args, **kwargs)
+
+    def locate(self, data):
+        raise NotImplementedError()
+
+
+class AbstractCellLocationProvider(AbstractLocationProvider):
+
```

Comment:
Maybe add a fake `__init__` or class attributes for the `self.` values used in this class, so it's obvious what a concrete implementation has to do? For example:
```
class AbstractCellLocationProvider(AbstractLocationProvider):

heka_client = None
models = ()
session = None
```